File: solutions/051.N-Queens/AC_dfs_n!.py
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dfs_n!.py
# Create Date: 2015-02-03 11:50:39
# Descripton:   

class Solution:
    def dfs(self, row):
        if row == self.n:
            self.res.append(self.cur[:])
            return
        for i in range(self.n):
            if self.col[i] != 0 or self.m_diag[row + i] != 0 or self.a_diag[row - i + self.n] != 0:
                continue
            self.col[i] = 1
            self.m_diag[row + i] = 1
            self.a_diag[row - i + self.n] = 1
            # So sad that str don't support item assignment in python
            self.cur[row] = self.cur[row][:i] + 'Q' + self.cur[row][i + 1:]
            self.dfs(row + 1)
            self.cur[row] = self.cur[row][:i] + '.' + self.cur[row][i + 1:]
            self.col[i] = 0
            self.m_diag[row + i] = 0
            self.a_diag[row - i + self.n] = 0

    # @return a list of lists of string
    def solveNQueens(self, n):
        self.res = []
        self.cur = ['.' * n] * n
        self.col = [0] * n
        self.m_diag = [0] * (2 * n)
        self.a_diag = [0] * (2 * n)
        self.n = n
        self.dfs(0)
        return self.res



s = Solution()
print s.solveNQueens(4)

File: solutions/077.Combinations/AC_dfs_n!.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_n!.cpp
*  Create Date: 2014-12-29 11:21:35
*  Descripton:  use dfs to generate combinations
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
private:
    void dfs(vector<vector<int> > &res, vector<int> &v,
            int cur, int rest, int n) {
        if (rest == 0) {
            res.push_back(v);
            return;
        }
        v.push_back(0);
        int len = v.size();
        for (int i = cur; i <= n - rest + 1; i++) {
            v[len - 1] = i;
            dfs(res, v, i + 1, rest - 1, n);
        }
        v.pop_back();
    }
public:
    vector<vector<int> > combine(int n, int k) {
        vector<vector<int> > res;
        if (n < k || !n || !k)
            return res;
        vector<int> v;
        dfs(res, v, 1, k, n);
        return res;
    }
};

int main() {
    int n, k;
    Solution s;
    while (cin >> n >> k) {
        vector<vector<int> > res = s.combine(n, k);
        for (auto &i : res) {
            for (auto &j : i)
                cout << j << ' ';
            cout << endl;
        }
    }
    return 0;
}

